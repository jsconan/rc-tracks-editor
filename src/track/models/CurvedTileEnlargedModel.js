/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { CURVED_TILE_ENLARGED_TYPE, TILE_DIRECTION_LEFT } from '../helpers';
import { TileModel } from './TileModel.js';
import { Vector2D } from '../../core/models';

/**
 * Represents a track tile for an enlarged curve.
 */
export class CurvedTileEnlargedModel extends TileModel {
    /**
     * Sets the size factor relative to a track section.
     * 1 means the tile fits 1 tile section in each direction.
     * @param {number} ratio - The size factor relative to a track section.
     * @returns {CurvedTileEnlargedModel} - Chains the instance.
     */
    setRatio(ratio) {
        this.ratio = Math.max(1, Math.floor(Math.abs(ratio)));
        return this;
    }

    /**
     * Computes the angle for rotating the tile to the expected direction.
     * @returns {number}
     */
    getDirectionAngle() {
        if (this.direction === TILE_DIRECTION_LEFT) {
            return Vector2D.RIGHT_ANGLE;
        }

        return 0;
    }

    /**
     * Computes the angle of the curve with respect to the ratio.
     * @returns {number}
     */
    getCurveAngle() {
        return Vector2D.RIGHT_ANGLE;
    }

    /**
     * Computes the length of the curve side if any.
     * @returns {number}
     */
    getCurveSide() {
        return this.length / 2;
    }

    /**
     * Computes the number of barrier chunks on each side with respect to the ratio.
     * @returns {number}
     */
    getSideBarrierChunks() {
        return (this.specs.barrierChunks * this.ratio) / 2;
    }

    /**
     * Computes the number of barrier chunks for the inner curve with respect to the ratio.
     * @returns {number}
     */
    getInnerBarrierChunks() {
        if (this.ratio <= 1) {
            return this.specs.barrierChunks / 2;
        }
        return this.specs.barrierChunks * this.ratio;
    }

    /**
     * Computes the number of barrier chunks for an outer curve with respect to the ratio.
     * @returns {number}
     */
    getOuterBarrierChunks() {
        return (this.specs.barrierChunks / 2) * this.ratio;
    }

    /**
     * Computes the coordinates of the center point.
     * @param {number} x - The X-coordinate of the tile.
     * @param {number} y - The Y-coordinate of the tile.
     * @param {number} angle - The rotation angle of the tile.
     * @returns {Vector2D}
     */
    getCenterCoord(x = 0, y = 0, angle = 0) {
        const start = new Vector2D(x, y);
        const center = start.addScalarY(this.specs.length * (this.ratio - 0.5));

        return center.rotateAround(angle, start);
    }

    /**
     * Computes the coordinates of the output point with respect to the tile direction.
     * @param {number} x - The X-coordinate of the tile.
     * @param {number} y - The Y-coordinate of the tile.
     * @param {number} angle - The rotation angle of the tile.
     * @returns {Vector2D}
     */
    getOutputCoord(x = 0, y = 0, angle = 0) {
        const start = new Vector2D(x, y);

        const radius = this.getInnerRadius() + this.specs.width / 2;
        const curveAngle = this.getCurveAngle();

        let center;
        if (this.direction === TILE_DIRECTION_LEFT) {
            center = start.addScalarX(radius);
        } else {
            center = start.subScalarX(radius);
        }

        return Vector2D.polar(radius, curveAngle, center).rotateAround(angle, start);
    }

    /**
     * Computes the angle of the output point with respect to the tile direction.
     * @param {number} angle - The rotation angle of the tile.
     * @returns {number}
     */
    getOutputAngle(angle = 0) {
        if (this.direction === TILE_DIRECTION_LEFT) {
            return Vector2D.degrees(angle - Vector2D.RIGHT_ANGLE);
        }

        return Vector2D.degrees(angle + Vector2D.RIGHT_ANGLE);
    }

    /**
     * Computes the coordinates of the edge point with respect to the tile direction.
     * @param {number} x - The X-coordinate of the tile.
     * @param {number} y - The Y-coordinate of the tile.
     * @param {number} angle - The rotation angle of the tile.
     * @returns {Vector2D}
     */
    getEdgeCoord(x = 0, y = 0, angle = 0) {
        const start = new Vector2D(x, y);

        const center = start.addScalarY(this.specs.length * (this.ratio - 0.5));
        const radius = this.getOuterRadius();
        let curveAngle;

        if (this.direction === TILE_DIRECTION_LEFT) {
            curveAngle = Vector2D.RIGHT_ANGLE + this.getCurveAngle() / 2;
        } else {
            curveAngle = Vector2D.RIGHT_ANGLE - this.getCurveAngle() / 2;
        }

        return Vector2D.polar(radius, curveAngle, center).rotateAround(angle, start);
    }
}

/**
 * The type of tile.
 * @constant {string} CurvedTileEnlargedModel.TYPE
 */
Object.defineProperty(CurvedTileEnlargedModel, 'TYPE', {
    value: CURVED_TILE_ENLARGED_TYPE,
    writable: false,
    enumerable: true,
    configurable: true
});
