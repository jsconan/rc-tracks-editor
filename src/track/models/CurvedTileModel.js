/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { CURVED_TILE_TYPE } from '../helpers';
import { TileModel } from './TileModel.js';
import { Vector2D } from '../../core/models';

/**
 * Represents a track tile for a curve.
 */
export class CurvedTileModel extends TileModel {
    /**
     * The actual length of the tile with respect to the ratio.
     * @type {number}
     */
    get length() {
        return this.specs.length;
    }

    /**
     * The actual width of the tile with respect to the ratio.
     * @type {number}
     */
    get width() {
        return this.specs.width;
    }

    /**
     * Computes the angle for rotating the tile to the left.
     * @returns {number}
     */
    getDirectionAngleLeft() {
        return 90 + (90 / this.ratio) * (Math.max(1, this.ratio) - 1);
    }

    /**
     * Computes the angle of the curve with respect to the ratio.
     * @returns {number}
     */
    getCurveAngle() {
        if (this.ratio < 1) {
            return 90 * this.ratio;
        }

        return 90 / this.ratio;
    }

    /**
     * Computes the coordinates of the center point.
     * @param {number} x - The X-coordinate of the tile.
     * @param {number} y - The Y-coordinate of the tile.
     * @param {number} angle - The rotation angle of the tile.
     * @returns {Vector2D}
     */
    getCenterCoord(x = 0, y = 0, angle = 0) {
        const start = new Vector2D(x, y);

        const curveAngle = this.getCurveAngle();
        const curveCenter = this.getCurveCenter().add(start);
        const radius = this.getInnerRadius() + this.specs.width / 2;

        const p1 = Vector2D.polar(radius, 0, curveCenter);
        const p2 = p1.addScalarY(10);
        const p3 = Vector2D.polar(radius, curveAngle, curveCenter);
        const p4 = p3.add(Vector2D.polar(10, curveAngle + 90));

        const center = Vector2D.intersect(p1, p2, p3, p4);

        return center.rotateAround(angle, start);
    }

    /**
     * Computes the coordinates of the output point when the tile is oriented to the right.
     * @param {number} x - The X-coordinate of the tile.
     * @param {number} y - The Y-coordinate of the tile.
     * @param {number} angle - The rotation angle of the tile.
     * @returns {Vector2D}
     */
    getOutputCoordRight(x = 0, y = 0, angle = 0) {
        const start = new Vector2D(x, y);

        const radius = this.getInnerRadius() + this.specs.width / 2;
        const curveAngle = this.getCurveAngle();
        const center = start.subScalarX(radius);

        return Vector2D.polar(radius, curveAngle, center).rotateAround(angle, start);
    }

    /**
     * Computes the coordinates of the output point when the tile is oriented to the left.
     * @param {number} x - The X-coordinate of the tile.
     * @param {number} y - The Y-coordinate of the tile.
     * @param {number} angle - The rotation angle of the tile.
     * @returns {Vector2D}
     */
    getOutputCoordLeft(x = 0, y = 0, angle = 0) {
        const start = new Vector2D(x, y);

        const radius = this.getInnerRadius() + this.specs.width / 2;
        const curveAngle = 180 - this.getCurveAngle();
        const center = start.addScalarX(radius);

        return Vector2D.polar(radius, curveAngle, center).rotateAround(angle, start);
    }

    /**
     * Computes the angle of the output point when the tile is oriented to the right.
     * @param {number} angle - The rotation angle of the tile.
     * @returns {number}
     */
    getOutputAngleRight(angle = 0) {
        return angle + this.getCurveAngle();
    }

    /**
     * Computes the angle of the output point when the tile is oriented to the left.
     * @param {number} angle - The rotation angle of the tile.
     * @returns {number}
     */
    getOutputAngleLeft(angle = 0) {
        return angle - this.getCurveAngle();
    }
}

/**
 * The type of tile.
 * @constant {string} CurvedTileModel.TYPE
 */
Object.defineProperty(CurvedTileModel, 'TYPE', {
    value: CURVED_TILE_TYPE,
    writable: false,
    enumerable: true,
    configurable: true
});
