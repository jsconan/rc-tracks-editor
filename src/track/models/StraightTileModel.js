/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { STRAIGHT_TILE_TYPE } from '../helpers';
import { TileModel } from './TileModel.js';
import { Vector2D } from '../../core/models';

/**
 * Represents a straight track tile.
 */
export class StraightTileModel extends TileModel {
    /**
     * The actual width of the tile with respect to the ratio.
     * @type {number}
     */
    get width() {
        return this.specs.width;
    }

    /**
     * Computes the angle for rotating the tile to the left.
     * @returns {number}
     */
    getDirectionAngleLeft() {
        return Vector2D.STRAIGHT_ANGLE;
    }
}

/**
 * The type of tile.
 * @constant {string} StraightTileModel.TYPE
 */
Object.defineProperty(StraightTileModel, 'TYPE', {
    value: STRAIGHT_TILE_TYPE,
    writable: false,
    enumerable: true,
    configurable: true
});
