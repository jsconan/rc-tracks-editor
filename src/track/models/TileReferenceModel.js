/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import {
    isDirectionValid,
    isTypeValid,
    CURVED_TILE_ENLARGED_TYPE,
    CURVED_TILE_TYPE,
    STRAIGHT_TILE_TYPE,
    TILE_DIRECTION_RIGHT
} from '../helpers';
import { uid } from '../../core/helpers';
import { CurvedTileEnlargedModel } from './CurvedTileEnlargedModel.js';
import { CurvedTileModel } from './CurvedTileModel.js';
import { StraightTileModel } from './StraightTileModel.js';
import { TileSpecifications } from './TileSpecifications';

/**
 * @type {object} - Maps the types of tile to their respective model.
 * @private
 */
const modelsMap = {
    [STRAIGHT_TILE_TYPE]: StraightTileModel,
    [CURVED_TILE_TYPE]: CurvedTileModel,
    [CURVED_TILE_ENLARGED_TYPE]: CurvedTileEnlargedModel
};

/**
 * Creates the tile model with respect to the given reference.
 * @param {TileReferenceModel} ref - The reference object.
 * @returns {TileModel} - Returns a tile model of the expected type.
 * @throws {TypeError} - If the given specifications object is not valid.
 * @throws {TypeError} - If the given direction is not valid.
 * @private
 */
function getModel(ref) {
    const Model = modelsMap[ref.type];
    return new Model(ref.specs, ref.direction, ref.ratio);
}

/**
 * Represents a reference to a tile model.
 */
export class TileReferenceModel {
    /**
     * Represents a reference to a tile model with the given specifications.
     * @param {TileSpecifications} specs - The specifications for the tiles.
     * @param {string} type - The type of referenced tile.
     * @param {string} direction - The direction of the tile, it can be either TILE_DIRECTION_RIGHT or TILE_DIRECTION_LEFT.
     * @param {number} ratio - The size ratio. Usually, it is included in the range [1-4].
     * @throws {TypeError} - If the given specifications object is not valid.
     * @throws {TypeError} - If the given type is not valid.
     * @throws {TypeError} - If the given direction is not valid.
     */
    constructor(specs, type = STRAIGHT_TILE_TYPE, direction = TILE_DIRECTION_RIGHT, ratio = 1) {
        this.setSpecs(specs);
        this.setType(type);
        this.setDirection(direction);
        this.setRatio(ratio);
        this.id = uid();
        this.model = getModel(this);
    }

    /**
     * The identifier for the model.
     * @type {string}
     */
    get modelId() {
        return this.model.id;
    }

    /**
     * Sets the specifications for the tiles.
     * @param {TileSpecifications} specs - The specifications for the tiles.
     * @returns {TileReferenceModel} - Chains the instance.
     * @throws {TypeError} - If the given specifications object is not valid.
     */
    setSpecs(specs) {
        if (!specs || !(specs instanceof TileSpecifications)) {
            throw new TypeError('A valid specifications object is needed!');
        }

        this.specs = specs;

        if (this.model) {
            this.model.setSpecs(specs);
        }

        return this;
    }

    /**
     * Sets the type of referenced tile.
     * @param {string} type - The type of referenced tile.
     * @returns {TileReferenceModel} - Chains the instance.
     * @throws {TypeError} - If the given type is not valid.
     */
    setType(type) {
        if (!isTypeValid(type)) {
            throw new TypeError('A valid type of tile is needed!');
        }

        this.type = type;

        if (this.model) {
            this.model = getModel(this);
        }

        return this;
    }

    /**
     * Sets the direction of the tile.
     * It can be either TileModel.DIRECTION_RIGHT or TileModel.DIRECTION_LEFT.
     * @param {string} direction - The direction of the tile.
     * @returns {TileReferenceModel} - Chains the instance.
     * @throws {TypeError} - If the given direction is not valid.
     */
    setDirection(direction) {
        if (!isDirectionValid(direction)) {
            throw new TypeError('A valid direction is needed!');
        }

        this.direction = direction;

        if (this.model) {
            this.model.setDirection(direction);
        }

        return this;
    }

    /**
     * Sets the size factor of the tile.
     * 1 means the tile fits 1 tile section in each direction.
     * @param {number} ratio - The size factor of the tile.
     * @returns {TileReferenceModel} - Chains the instance.
     */
    setRatio(ratio) {
        this.ratio = Math.abs(ratio || 1);

        if (this.model) {
            this.model.setRatio(ratio);
        }

        return this;
    }

    /**
     * Exports the model to an object.
     * @returns {tileExport} - An object representation of the model.
     */
    export() {
        const { type, direction, ratio } = this;
        return { type, direction, ratio };
    }

    /**
     * Computes the coordinates of the tile.
     * @param {number} x - The X-coordinate of the tile.
     * @param {number} y - The Y-coordinate of the tile.
     * @param {number} angle - The rotation angle of the tile.
     * @returns {tileCoord} - The computed coordinates for rendering the tile.
     * @throws {TypeError} - If the given specifications object is not valid.
     */
    build(x = 0, y = 0, angle = 0) {
        const { id, type, model } = this;
        const outputAngle = model.getOutputAngle(angle);
        const outputCoord = model.getOutputCoord(x, y, angle);
        const centerCoord = model.getCenterCoord(x, y, angle);

        return { id, type, x, y, angle, centerCoord, outputCoord, outputAngle, model };
    }
}

/**
 * @typedef {object} tileCoord - Represents a tile ready to be rendered.
 * @property {string} id - The identifier of the tile.
 * @property {string} type - The type of tile.
 * @property {number} x - The left coordinate of the tile.
 * @property {number} y - The top coordinate of the tile.
 * @property {number} angle - The rotation angle of the tile.
 * @property {Vector2D} centerCoord - The coordinates of the center of the tile.
 * @property {Vector2D} outputCoord - The coordinates of the output of the tile.
 * @property {number} outputAngle - The rotation angle at the output of the tile.
 * @property {TileModel} model - The tile model with respect to its type.
 */

/**
 * @typedef {object} tileExport - Represents an exported tile reference.
 * @property {string} type - The type of tile.
 * @property {string} direction - The direction of the tile.
 * @property {number} ratio - The size ratio of the tile.
 */

/**
 * @typedef {import('./TileModel').TileModel} TileModel
 */

/**
 * @typedef {import('../../core/models/Vector2D').Vector2D} Vector2D
 */
