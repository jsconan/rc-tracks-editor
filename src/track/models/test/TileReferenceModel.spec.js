/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import {
    CURVED_TILE_ENLARGED_TYPE,
    CURVED_TILE_TYPE,
    STRAIGHT_TILE_TYPE,
    TILE_DIRECTION_LEFT,
    TILE_DIRECTION_RIGHT
} from '../../helpers';
import { CurvedTileModel } from '../CurvedTileModel.js';
import { CurvedTileEnlargedModel } from '../CurvedTileEnlargedModel.js';
import { StraightTileModel } from '../StraightTileModel.js';
import { TileReferenceModel } from '../TileReferenceModel.js';
import { TileSpecifications } from '../TileSpecifications.js';

const laneWidth = 80;
const barrierWidth = 5;
const barrierChunks = 4;
const specs = new TileSpecifications(laneWidth, barrierWidth, barrierChunks);

describe('TileReferenceModel', () => {
    it('is a class', () => {
        expect(TileReferenceModel).toEqual(expect.any(Function));
    });

    describe('throws error', () => {
        it('when trying to create an instance with an invalid specifications object', () => {
            // @ts-expect-error
            expect(() => new TileReferenceModel({})).toThrow('A valid specifications object is needed!');
        });

        it('when trying to create an instance with an invalid type', () => {
            expect(() => new TileReferenceModel(specs, '')).toThrow('A valid type of tile is needed!');
        });

        it('when trying to create an instance with an invalid direction', () => {
            expect(() => new TileReferenceModel(specs, STRAIGHT_TILE_TYPE, '')).toThrow('A valid direction is needed!');
        });

        it('when trying to set an invalid specifications object', () => {
            const track = new TileReferenceModel(specs);
            // @ts-expect-error
            expect(() => track.setSpecs({})).toThrow('A valid specifications object is needed!');
        });

        it('when trying to set an invalid type', () => {
            const ref = new TileReferenceModel(specs, STRAIGHT_TILE_TYPE);
            expect(() => ref.setType('')).toThrow('A valid type of tile is needed!');
        });

        it('when trying to set an invalid direction', () => {
            const ref = new TileReferenceModel(specs, STRAIGHT_TILE_TYPE);
            expect(() => ref.setDirection('')).toThrow('A valid direction is needed!');
        });
    });

    describe('can build a reference to a tile', () => {
        it('with the given specifications', () => {
            const ref = new TileReferenceModel(specs);

            expect(ref).toBeInstanceOf(TileReferenceModel);
            expect(ref).toMatchSnapshot();
            expect(ref.modelId).toBe(`${STRAIGHT_TILE_TYPE}-1`);
        });

        it('with the given type', () => {
            const ref = new TileReferenceModel(specs, STRAIGHT_TILE_TYPE);

            expect(ref).toBeInstanceOf(TileReferenceModel);
            expect(ref).toMatchSnapshot();
            expect(ref.modelId).toBe(`${STRAIGHT_TILE_TYPE}-1`);
        });

        it('with the given direction', () => {
            const ref = new TileReferenceModel(specs, STRAIGHT_TILE_TYPE, TILE_DIRECTION_LEFT);

            expect(ref).toBeInstanceOf(TileReferenceModel);
            expect(ref).toMatchSnapshot();
            expect(ref.modelId).toBe(`${STRAIGHT_TILE_TYPE}-1`);
        });

        it('with the given ratio', () => {
            const ref = new TileReferenceModel(specs, STRAIGHT_TILE_TYPE, TILE_DIRECTION_LEFT, 2);

            expect(ref).toBeInstanceOf(TileReferenceModel);
            expect(ref).toMatchSnapshot();
            expect(ref.modelId).toBe(`${STRAIGHT_TILE_TYPE}-2`);
        });
    });

    describe('can build the model for', () => {
        it('a straight tile', () => {
            const ref = new TileReferenceModel(specs, STRAIGHT_TILE_TYPE);

            expect(ref.model).toBeInstanceOf(StraightTileModel);
            expect(ref.model).toMatchSnapshot();
        });

        it('a curved tile', () => {
            const ref = new TileReferenceModel(specs, CURVED_TILE_TYPE);

            expect(ref.model).toBeInstanceOf(CurvedTileModel);
            expect(ref.model).toMatchSnapshot();
        });

        it('a curved tile enlarged', () => {
            const ref = new TileReferenceModel(specs, CURVED_TILE_ENLARGED_TYPE);

            expect(ref.model).toBeInstanceOf(CurvedTileEnlargedModel);
            expect(ref.model).toMatchSnapshot();
        });
    });

    describe('can set', () => {
        it('the specifications of the tile', () => {
            const ref = new TileReferenceModel(specs);
            const newSpecs = new TileSpecifications(10, 1, 2);

            expect(ref.specs).toBeInstanceOf(TileSpecifications);
            expect(ref.specs).not.toBe(newSpecs);
            expect(ref.model.specs).not.toBe(newSpecs);
            expect(ref.setSpecs(newSpecs)).toBe(ref);
            expect(ref.specs).toBe(newSpecs);
            expect(ref.model.specs).toBe(newSpecs);
        });

        it('the type of tile', () => {
            const ref = new TileReferenceModel(specs);

            expect(ref.type).toBe(STRAIGHT_TILE_TYPE);
            expect(ref.model).toBeInstanceOf(StraightTileModel);
            expect(ref.setType(CURVED_TILE_TYPE)).toBe(ref);
            expect(ref.type).toBe(CURVED_TILE_TYPE);
            expect(ref.model).toBeInstanceOf(CurvedTileModel);
            expect(ref.modelId).toBe(`${CURVED_TILE_TYPE}-1`);
        });

        it('the direction of the tile', () => {
            const ref = new TileReferenceModel(specs);

            expect(ref.direction).toBe(TILE_DIRECTION_RIGHT);
            expect(ref.model.direction).toBe(TILE_DIRECTION_RIGHT);
            expect(ref.setDirection(TILE_DIRECTION_LEFT)).toBe(ref);
            expect(ref.direction).toBe(TILE_DIRECTION_LEFT);
            expect(ref.model.direction).toBe(TILE_DIRECTION_LEFT);
            expect(ref.modelId).toBe(`${STRAIGHT_TILE_TYPE}-1`);
        });

        it('the ratio of the tile', () => {
            const ref = new TileReferenceModel(specs);

            expect(ref.ratio).toBe(1);
            expect(ref.model.ratio).toBe(1);
            expect(ref.setRatio(2)).toBe(ref);
            expect(ref.ratio).toBe(2);
            expect(ref.model.ratio).toBe(2);
            expect(ref.modelId).toBe(`${STRAIGHT_TILE_TYPE}-2`);
            expect(ref.setRatio(-1).ratio).toBe(1);
            expect(ref.setRatio(0).ratio).toBe(1);
            expect(ref.setRatio(0.5).ratio).toBe(0.5);
        });
    });

    it.each([STRAIGHT_TILE_TYPE, CURVED_TILE_TYPE, CURVED_TILE_ENLARGED_TYPE])(
        'can export to an object from a tile type %s',
        type => {
            const track = new TileReferenceModel(specs, type);

            expect(track.export()).toMatchSnapshot();
        }
    );

    describe('can compute the coordinates of the tile for', () => {
        it.each([
            [void 0, void 0, void 0],
            [100, 100, 45]
        ])('a straight tile at coordinates [%s, %s] and angle %s', (x, y, angle) => {
            const ref = new TileReferenceModel(specs, STRAIGHT_TILE_TYPE);
            expect(ref.build(x, y, angle)).toMatchSnapshot();
        });

        it.each([
            [void 0, void 0, void 0],
            [100, 100, 45]
        ])('a curved tile at coordinates [%s, %s] and angle %s', (x, y, angle) => {
            const ref = new TileReferenceModel(specs, CURVED_TILE_TYPE);
            expect(ref.build(x, y, angle)).toMatchSnapshot();
        });

        it.each([
            [void 0, void 0, void 0],
            [100, 100, 45]
        ])('a curved tile enlarged at coordinates [%s, %s] and angle %s', (x, y, angle) => {
            const ref = new TileReferenceModel(specs, CURVED_TILE_ENLARGED_TYPE);
            expect(ref.build(x, y, angle)).toMatchSnapshot();
        });
    });
});
