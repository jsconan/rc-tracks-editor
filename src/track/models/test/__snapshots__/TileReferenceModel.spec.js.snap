// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TileReferenceModel can build a reference to a tile with the default specifications 1`] = `
TileReferenceModel {
  "direction": "right",
  "id": "id-6",
  "ratio": 1,
  "type": "straight-tile",
}
`;

exports[`TileReferenceModel can build a reference to a tile with the given direction 1`] = `
TileReferenceModel {
  "direction": "left",
  "id": "id-8",
  "ratio": 1,
  "type": "straight-tile",
}
`;

exports[`TileReferenceModel can build a reference to a tile with the given ratio 1`] = `
TileReferenceModel {
  "direction": "left",
  "id": "id-9",
  "ratio": 2,
  "type": "straight-tile",
}
`;

exports[`TileReferenceModel can build a reference to a tile with the given type 1`] = `
TileReferenceModel {
  "direction": "right",
  "id": "id-7",
  "ratio": 1,
  "type": "straight-tile",
}
`;

exports[`TileReferenceModel can build the model for a curved tile 1`] = `
CurvedTileModel {
  "direction": "right",
  "ratio": 1,
  "specs": TileSpecifications {
    "barrierChunks": 4,
    "barrierWidth": 5,
    "laneWidth": 80,
  },
}
`;

exports[`TileReferenceModel can build the model for a curved tile enlarged 1`] = `
CurvedTileEnlargedModel {
  "direction": "right",
  "ratio": 1,
  "specs": TileSpecifications {
    "barrierChunks": 4,
    "barrierWidth": 5,
    "laneWidth": 80,
  },
}
`;

exports[`TileReferenceModel can build the model for a straight tile 1`] = `
StraightTileModel {
  "direction": "right",
  "ratio": 1,
  "specs": TileSpecifications {
    "barrierChunks": 4,
    "barrierWidth": 5,
    "laneWidth": 80,
  },
}
`;

exports[`TileReferenceModel can compute the coordinates of the tile for a curved tile at coordinates [100, 100] and angle 45 1`] = `
Object {
  "angle": 45,
  "centerCoord": Vector2D {
    "x": 61.10912703473989,
    "y": 138.89087296526012,
  },
  "id": "id-19",
  "model": CurvedTileModel {
    "direction": "right",
    "ratio": 1,
    "specs": TileSpecifications {
      "barrierChunks": 4,
      "barrierWidth": 5,
      "laneWidth": 80,
    },
  },
  "outputAngle": 135,
  "outputCoord": Vector2D {
    "x": 22.21825406947977,
    "y": 100,
  },
  "type": "curved-tile",
  "x": 100,
  "y": 100,
}
`;

exports[`TileReferenceModel can compute the coordinates of the tile for a curved tile at coordinates [undefined, undefined] and angle undefined 1`] = `
Object {
  "angle": 0,
  "centerCoord": Vector2D {
    "x": 0,
    "y": 55,
  },
  "id": "id-18",
  "model": CurvedTileModel {
    "direction": "right",
    "ratio": 1,
    "specs": TileSpecifications {
      "barrierChunks": 4,
      "barrierWidth": 5,
      "laneWidth": 80,
    },
  },
  "outputAngle": 90,
  "outputCoord": Vector2D {
    "x": -55,
    "y": 55,
  },
  "type": "curved-tile",
  "x": 0,
  "y": 0,
}
`;

exports[`TileReferenceModel can compute the coordinates of the tile for a curved tile enlarged at coordinates [100, 100] and angle 45 1`] = `
Object {
  "angle": 45,
  "centerCoord": Vector2D {
    "x": 61.10912703473989,
    "y": 138.89087296526012,
  },
  "id": "id-21",
  "model": CurvedTileEnlargedModel {
    "direction": "right",
    "ratio": 1,
    "specs": TileSpecifications {
      "barrierChunks": 4,
      "barrierWidth": 5,
      "laneWidth": 80,
    },
  },
  "outputAngle": 135,
  "outputCoord": Vector2D {
    "x": 22.21825406947977,
    "y": 100,
  },
  "type": "curved-tile-enlarged",
  "x": 100,
  "y": 100,
}
`;

exports[`TileReferenceModel can compute the coordinates of the tile for a curved tile enlarged at coordinates [undefined, undefined] and angle undefined 1`] = `
Object {
  "angle": 0,
  "centerCoord": Vector2D {
    "x": 0,
    "y": 55,
  },
  "id": "id-20",
  "model": CurvedTileEnlargedModel {
    "direction": "right",
    "ratio": 1,
    "specs": TileSpecifications {
      "barrierChunks": 4,
      "barrierWidth": 5,
      "laneWidth": 80,
    },
  },
  "outputAngle": 90,
  "outputCoord": Vector2D {
    "x": -55,
    "y": 55,
  },
  "type": "curved-tile-enlarged",
  "x": 0,
  "y": 0,
}
`;

exports[`TileReferenceModel can compute the coordinates of the tile for a straight tile at coordinates [100, 100] and angle 45 1`] = `
Object {
  "angle": 45,
  "centerCoord": Vector2D {
    "x": 61.10912703473989,
    "y": 138.89087296526012,
  },
  "id": "id-17",
  "model": StraightTileModel {
    "direction": "right",
    "ratio": 1,
    "specs": TileSpecifications {
      "barrierChunks": 4,
      "barrierWidth": 5,
      "laneWidth": 80,
    },
  },
  "outputAngle": 45,
  "outputCoord": Vector2D {
    "x": 22.218254069479784,
    "y": 177.78174593052023,
  },
  "type": "straight-tile",
  "x": 100,
  "y": 100,
}
`;

exports[`TileReferenceModel can compute the coordinates of the tile for a straight tile at coordinates [undefined, undefined] and angle undefined 1`] = `
Object {
  "angle": 0,
  "centerCoord": Vector2D {
    "x": 0,
    "y": 55,
  },
  "id": "id-16",
  "model": StraightTileModel {
    "direction": "right",
    "ratio": 1,
    "specs": TileSpecifications {
      "barrierChunks": 4,
      "barrierWidth": 5,
      "laneWidth": 80,
    },
  },
  "outputAngle": 0,
  "outputCoord": Vector2D {
    "x": 0,
    "y": 110,
  },
  "type": "straight-tile",
  "x": 0,
  "y": 0,
}
`;
